src/provider/mod.rs

```diff
@@ -0,0 +1,192 @@
+use std::convert::TryFrom;
+use std::fs;
+use std::path::PathBuf;
+use std::process::Child;
+
+use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
+use kube::{Api, Client};
+use kubelet::backoff::ExponentialBackoffStrategy;
+use kubelet::log::Sender;
+use kubelet::node::Builder;
+use kubelet::pod::Pod;
+use kubelet::provider::Provider;
+use log::{debug, error};
+
+use crate::provider::error::StackableError;
+use crate::provider::error::StackableError::{CrdMissing, KubeError, PodValidationError};
+use crate::provider::repository::package::Package;
+use crate::provider::states::downloading::Downloading;
+use crate::provider::states::terminated::Terminated;
+use kube::error::ErrorResponse;
+
+pub struct StackableProvider {
+    client: Client,
+    parcel_directory: PathBuf,
+    config_directory: PathBuf,
+    log_directory: PathBuf,
+}
+
+pub const CRDS: &'static [&'static str] = &["repositories.stable.stackable.de"];
+
+mod error;
+mod repository;
+mod states;
+
+pub struct PodState {
+    client: Client,
+    parcel_directory: PathBuf,
+    download_directory: PathBuf,
+    config_directory: PathBuf,
+    log_directory: PathBuf,
+    package_download_backoff_strategy: ExponentialBackoffStrategy,
+    service_name: String,
+    package: Package,
+    process_handle: Option<Child>,
+}
+
+impl PodState {
+    pub fn get_service_config_directory(&self) -> PathBuf {
+        self.config_directory.join(&self.service_name)
+    }
+
+    pub fn get_service_package_directory(&self) -> PathBuf {
+        self.parcel_directory
+            .join(&self.package.get_directory_name())
+    }
+
+    pub fn get_service_log_directory(&self) -> PathBuf {
+        self.log_directory.join(&self.service_name)
+    }
+}
+
+impl StackableProvider {
+    pub async fn new(
+        client: Client,
+        parcel_directory: PathBuf,
+        config_directory: PathBuf,
+        log_directory: PathBuf,
+    ) -> Result<Self, StackableError> {
+        let provider = StackableProvider {
+            client,
+            parcel_directory,
+            config_directory,
+            log_directory,
+        };
+        let missing_crds = provider.check_crds().await?;
+        if missing_crds.is_empty() {
+            debug!("All required CRDS present!");
+            return Ok(provider);
+        } else {
+            debug!("Missing required CDRS");
+            return Err(CrdMissing { missing_crds });
+        }
```

Comment:
return can be lifted out of the if (I'm only saying "can" if you prefer this style I'm fine as well)